[skip ci] Hack: Fix get_ci_vm.sh w/ gcloud ssh/scp

cevich opened the pull request:
Previously, using the ssh command directly required obtaining the
external IP of the VM and was then subject to the local configuration.
If the local configuration and/or ssh keys are incorrect, these commands
would fail, preventing automatic setup of the VM.

Fix this by using the gcloud ssh and scp wrappers.  Unfortunately rsync
couldn't be made to work in this situation, so use a tarball to transfer
the local repository to the VM.  Lastly, execute `setup_environment.sh`
script, then drop the caller into a bash shell sitting in the remote
`$GOSRC` directory.

Signed-off-by: Chris Evich <cevich@redhat.com>

cevich commented:
Background: ``hack/get_ci_vm.sh`` allows anyone with access to the libpod GCE project to easily create and access VM's exactly matching what's used in cirrus-ci.  For reproducible test problems, doing it this way is **FAR** easier/faster than trying to grab an executing VM manually.

When you have 5-minutes, @rhatdan @baude @mheon @giuseppe would one of you mind giving this script a "I've never done this before" try?  

I want to make sure it works for ``!me``.   Usage should be fairly self-explanatory.  Otherwise ping me if you're able to break it.

rhatdan commented:
Tried it out, but I need gcloud installed.

rhatdan commented on hack/get_ci_vm.sh:
```diff
@@ -13,16 +13,42 @@ DISK="200"
 GOSRC="/var/tmp/go/src/github.com/containers/libpod"
 
-PGCLOUD="sudo podman run -it --rm -e AS_ID=$UID -e AS_USER=$USER -v /home/$USER:$HOME:z quay.io/cevich/gcloud_centos:latest"
+PGCLOUD="sudo podman run -it --rm -e AS_ID=$UID -e AS_USER=$USER -v /home/$USER:$HOME:z -v /tmp:/tmp:ro quay.io/cevich/gcloud_centos:latest"
```
The `:z` on this line is definitely WRONG.  This will cause your entire homedir to be mislabeled.

Remove the :z and switch to using --security-option lable=disabled

To turn off SELinux separation.

cevich replied:
Oooohhh, ok that sounds like a much better idea.  Thanks!

rhatdan commented:
My script uses 
```
PGCLOUD="sudo podman run -it --rm -e AS_ID=$UID -e AS_USER=$USER -v /home/$USER:$HOME --security-opt label=disable -v /tmp:/tmp:ro quay.io/cevich/gcloud_centos:latest"
```

giuseppe commented:
after tweaking the :z to not change my home dir labels:
```
$ hack/get_ci_vm.sh ubuntu-18-libpod-86d821e

WARNING: This will not work without local sudo access to run podman,
         and prior authorization to use the libpod GCP project. Also,
         possession of the proper ssh private key is required.

Packing up repository into a tarball gscrivano-twidling-ubuntu-18-libpod-86d821e.
+ tar cjf /tmp/get_ci_vm.sh_ClB3r.tar.bz2 --warning=no-file-changed . &
<backgrounded>

Trying to creating a VM named gscrivano-twidling-ubuntu-18-libpod-86d821e (errors ignored).
+ sudo podman run -it --rm -e AS_ID=1000 -e AS_USER=gscrivano -v /home/gscrivano:/home/gscrivano -v /tmp:/tmp:ro quay.io/cevich/gcloud_centos:latest compute instances create --zone=us-central1-a --image=ubuntu-18-libpod-86d821e --custom-cpu=2 --custom-memory=4Gb --boot-disk-size=200 --labels=in-use-by=gscrivano gscrivano-twidling-ubuntu-18-libpod-86d821e
+ exec sudo --set-home --user gscrivano --login --stdin /usr/bin/gcloud compute instances create --zone=us-central1-a --image=ubuntu-18-libpod-86d821e --custom-cpu=2 --custom-memory=4Gb --boot-disk-size=200 --labels=in-use-by=gscrivano gscrivano-twidling-ubuntu-18-libpod-86d821e
ERROR: (gcloud.compute.instances.create) Could not fetch resource:
 - Invalid value for field 'resource.disks[0].initializeParams.sourceImage': 'https://www.googleapis.com/compute/v1/projects/api-project-281423873376/global/images/ubuntu-18-libpod-86d821e'. The referenced image resource cannot be found.


Removing and re-creating /var/tmp/go/src/github.com/containers/libpod on gscrivano-twidling-ubuntu-18-libpod-86d821e.
+ sudo podman run -it --rm -e AS_ID=1000 -e AS_USER=gscrivano -v /home/gscrivano:/home/gscrivano -v /tmp:/tmp:ro quay.io/cevich/gcloud_centos:latest compute ssh root@gscrivano-twidling-ubuntu-18-libpod-86d821e --command rm -rf /var/tmp/go/src/github.com/containers/libpod
+ exec sudo --set-home --user gscrivano --login --stdin /usr/bin/gcloud compute ssh root@gscrivano-twidling-ubuntu-18-libpod-86d821e --command 'rm -rf /var/tmp/go/src/github.com/containers/libpod'
ERROR: (gcloud.compute.ssh) Could not fetch resource:
 - The resource 'projects/api-project-281423873376/zones/europe-west1-b/instances/gscrivano-twidling-ubuntu-18-libpod-86d821e' was not found




Deleting gscrivano-twidling-ubuntu-18-libpod-86d821e (Might take a minute or two)
+ sudo podman run -it --rm -e AS_ID=1000 -e AS_USER=gscrivano -v /home/gscrivano:/home/gscrivano -v /tmp:/tmp:ro quay.io/cevich/gcloud_centos:latest compute instances delete --zone us-central1-a --delete-disks=all gscrivano-twidling-ubuntu-18-libpod-86d821e
+ exec sudo --set-home --user gscrivano --login --stdin /usr/bin/gcloud compute instances delete --zone us-central1-a --delete-disks=all gscrivano-twidling-ubuntu-18-libpod-86d821e
The following instances will be deleted. Any attached disks configured
 to be auto-deleted will be deleted unless they are attached to any 
other instances or the `--keep-disks` flag is given and specifies them
 for keeping. Deleting a disk is irreversible and any data on the disk
 will be lost.
 - [gscrivano-twidling-ubuntu-18-libpod-86d821e] in [us-central1-a]

Do you want to continue (Y/n)?  y

ERROR: (gcloud.compute.instances.delete) Failed to fetch some instances:
 - The resource 'projects/api-project-281423873376/zones/us-central1-a/instances/gscrivano-twidling-ubuntu-18-libpod-86d821e' was not found
```

rhatdan commented:
@giuseppe you also need the --security-opt label=disabled

giuseppe commented:
please include these changes:

```
diff --git a/hack/get_ci_vm.sh b/hack/get_ci_vm.sh
index bb65c772..9c334969 100755
--- a/hack/get_ci_vm.sh
+++ b/hack/get_ci_vm.sh
@@ -13,7 +13,7 @@ DISK="200"
 PROJECT="libpod-218412"
 GOSRC="/var/tmp/go/src/github.com/containers/libpod"
 
-PGCLOUD="sudo podman run -it --rm -e AS_ID=$UID -e AS_USER=$USER -v /home/$USER:$HOME:z -v /tmp:/tmp:ro quay.io/cevich/gcloud_centos:latest"
+PGCLOUD="sudo podman run -it --rm -e AS_ID=$UID -e AS_USER=$USER -v /home/$USER:$HOME --security-opt label=disable -v /tmp:/tmp:ro quay.io/cevich/gcloud_centos:latest"
 CREATE_CMD="$PGCLOUD compute instances create --zone=$ZONE --image=$1 --custom-cpu=$CPUS --custom-memory=$MEMORY --boot-disk-size=$DISK --labels=in-use-by=$USER $VMNAME"
 SSH_CMD="$PGCLOUD compute ssh root@$VMNAME"
 SCP_CMD="$PGCLOUD compute scp"
@@ -44,8 +44,6 @@ cleanup() {
     showrun $CLEANUP_CMD  # prompts for Yes/No
     rm -f "$TEMPFILE"
 }
-trap cleanup EXIT
-
 
 ##### main
 
@@ -56,6 +54,8 @@ then
     exit 1
 fi
 
+trap cleanup EXIT
+
 echo -e "\n${YEL}WARNING: This will not work without local sudo access to run podman,${NOR}"
 echo -e "         ${YEL}and prior authorization to use the libpod GCP project. Also,${NOR}"
 echo -e "         ${YEL}possession of the proper ssh private key is required.${NOR}"
```

I  am still getting some errors like:

```
ERROR: (gcloud.compute.instances.create) Could not fetch machine type:
 - Required 'compute.machineTypes.get' permission for 'projects/api-project-281423873376/zones/us-central1-a/machineTypes/custom-2-4096'
```

but I guess it depends on my configuration not being correct

cevich commented:
...okay, fixed SELinux thing, added more robust configuration testing, and added support for initializing with source or package-based libpod dependencies (making this PR depend on #1936).  Tested with existing and non-existing `$HOME/.config/gcloud` directory.  I'll come back here once #1936 is more concrete.

cevich commented:
...fixed (removed) dependency on #1936 but needs more testing.

cevich commented:
@baude you're the next in line...mind playing guinea pig with this script?  Nothing to install, it should 'just work'.  I believe I got most of the kinks out that Dan/Giuseppe found :smile:

cevich commented:
Tested and fixed removal of dependency on #1936.  This script will work in both cases now.

cevich commented:
/approve

cevich commented:
Aww, can't /approve my own PR, shucks :fox_face:

mheon commented:
/approve

rhatdan commented:
LGTM
@giuseppe PTAL

cevich commented:
(rebased to latest master and `--force` pushed)

cevich commented:
@giuseppe mind taking a quick second-poke at this?  I've been using it over the past weeks and found it working well.  However, I'm most curious if the "not cevich" user experience is positive as well :D

giuseppe commented:
I tried it and it works for me.

/lgtm

cevich commented:
Thanks @giuseppe now I have something to demo :grin:
